moshi/src/main/java/com/squareup/moshi/Moshi.java

```diff
@@ -133,20 +148,46 @@
           synchronized (adapterCache) {
             adapterCache.put(cacheKey, result);
           }
+          // Remove the field name only when we succeed in creating the adapter,
+          // so we have the full stack if we need to catch an exception at the top level.
+          if (fieldName != null) {
+            fieldNames.remove(fieldNames.size() - 1);
+          }
           return result;
         }
       }
+    } catch (RuntimeException e) {
+      if (lastIndex == 0) { // Rewrite the exception at the top level.
+        e = errorWithFields(type, fieldNames, e);
+      }
+      throw e;
     } finally {
-      deferredAdapters.remove(deferredAdapters.size() - 1);
+      deferredAdapters.remove(lastIndex);
       if (deferredAdapters.isEmpty()) {
```

Comment:
We should combine the list with the deferred adapters. Maybe give a deferred adapter a name?

Reply:
Do you mean make a DeferredAdapter subclass to hold onto the field stack?